Add LINQ Shuffle

stephentoub opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/111221

stephentoub committed:
Add LINQ Shuffle

LeaFrock commented on src/libraries/System.Linq/src/System/Linq/Shuffle.cs:
```diff
@@ -0,0 +1,91 @@
+                }
+                else if (state == 1)
+                {
+                    TSource[] buffer = _source.ToArray();
```
So `Shuffle` will always call `source.ToArray` to allocate an array, and never change the source elements' positions. Would it be better add some remarks in the API doc?

huoyaoyuan replied:
This is how all LINQ methods work.

LeaFrock commented on src/libraries/System.Linq/src/System/Linq/Shuffle.cs:
```diff
@@ -0,0 +1,91 @@
+                    TSource[] buffer = _source.ToArray();
+                    if (buffer.Length != 0)
+                    {
+                        Random.Shared.Shuffle(buffer);
```
Is it possible to deal with a specific case `Shuffle().Take(1)`? When randomly picking one element only, shuffling all elements seems to be a waste.

stephentoub replied:
It's possible, but it requires either adding another separate iterator impl or adding the additional take count to this one and factoring it into all the implementations. It's also much more complicated for counts larger than 1. That's all why I skipped it initially.

LeaFrock replied:
> That's all why I skipped it initially.

Got it. I think `Shuffle().Take(N)` will become a relatively hot requirement after releasing this API. It is indeed a performance detail if a separate iterator(`TakeShuffleIterator<TSource>`) can reduce the element swap times with `O(N)` instead of `O(Length)`. I see your another commit adding it. That's wonderful 👍 !

eiriktsarpalis commented on src/libraries/System.Linq/src/System/Linq/Shuffle.cs:
```diff
@@ -0,0 +1,91 @@
+                return [];
+            }
+
+            if (source is ShuffleIterator<TSource> shuffleIterator)
```
Seems like an obvious optimization, but is it realistically common enough to force another type check?

eiriktsarpalis commented on src/libraries/System.Linq/tests/Shuffler.cs:
```diff
@@ -14,6 +14,6 @@ public static T[] Shuffle<T>(T[] array)
         }
 
-        public static IEnumerable<T> Shuffle<T>(this IEnumerable<T> source) => Shuffle(source.ToArray());
+        public static IEnumerable<T> Shuffle<T>(IEnumerable<T> source) => Shuffle(source.ToArray());
```
Can't we just get rid of this now?

stephentoub replied:
Possibly, but the tests were currently deterministic, and removing this and changing all call sites to use the new one would make them not.

eiriktsarpalis replied:
Seems like another argument in favor of a deterministic overload?

stephentoub commented on src/libraries/System.Linq/src/System/Linq/Shuffle.cs:
```diff
@@ -0,0 +1,91 @@
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+
+            if (IsEmptyArray(source))
```
> Should we also add a bespoke implementation targeting IList<T> directly?

What would it do differently?

eiriktsarpalis replied:
I was thinking something along the following lines:
```C#
static IEnumerable<T> Shuffle<T>(IList<T> source)
{
    int n = source.Count;
    int[]? pooledBuffer = null;

    Span<int> permutationMap = n <= 64
        ? stackalloc int[64]
        : (pooledBuffer = ArrayPool<int>.Shared.Rent(n));

    permutationMap = permutationMap.Slice(n);
    PopulatePermutationMap(permutationMap); // Similar to Random.Shuffle

    try
    {
        for (int i = 0; i < n; i++)
        {
            yield return source[permutationMap[i]];
        }
    }
    finally
    {
        if (pooledBuffer != null)
        {
            ArrayPool<int>.Shared.Return(pooledBuffer);
        }
    }
}
```

eiriktsarpalis replied:
Minus the stackalloc which can't be used on iterator methods (Weirdly enough Intellisense didn't light up with an error when I was playing with the idea)

stephentoub replied:
I'm a little hesitant to use ArrayPool like that in an iterator across enumeration; it means leaking a buffer if the enumerator isn't disposed, for example. Do we use ArrayPool like that anywhere else in enumeration in LINQ? I thought we only used it in ToArray/ToList/etc.

eiriktsarpalis replied:
In hindsight I think it's fine to keep `ToArray` for everything. I was hoping I could pin this down to a simpler solution, but I couldn't find anything not forcing either $O(n)$ space, additional iterations through the collection like above, or $O(n^2)$ time.

stephentoub committed:
Merge branch 'main' into addshuffle

stephentoub committed:
Remove bespoke Shuffler

stephentoub committed:
Remove check for ShuffleIterator

stephentoub committed:
Optimize Shuffle().Take() with reservoir sampling

eiriktsarpalis commented on src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs:
```diff
@@ -0,0 +1,272 @@
+                    using IEnumerator<TSource> e = source.GetEnumerator();
+                    if (e.MoveNext())
+                    {
+                        reservoir = [e.Current];
```
Shouldn't the reservoir also be presized to `takeCount` here?

stephentoub replied:
We don't know how many elements will actually be in the source. It could be much less than the takeCount.

LeaFrock commented on src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs:
```diff
@@ -68,6 +69,204 @@ public override int GetCount(bool onlyIfCheap) =>
+                    int listCount = list.Count;
+                    Debug.Assert(listCount > takeCount, "Known listCount <= takeCount should have been handled by Iterator.Take override");
+
+                    reservoir = new(takeCount);
```
It's so special if `takeCount` is `1` so that the list allocated here (even though the capacity has been set) might be removed in theory. The cost is a `bool` branch and more code lines.

I don't know how to judge the balance 🤣 . It seems that randomly picking a single element is more often than other count numbers.

stephentoub replied:
> It seems that randomly picking a single element is more often than other count numbers.

How so?

I do not think it's worth spending even more code optimizing that case until it's proven to really be necessary. It would non-trivially complicate things, all to avoid an extra allocation, when we're already allocating for the Shuffle enumerable and the Take enumerable and whatever the input to the Shuffle was.

LeaFrock replied:
> until it's proven to really be necessary

Got it.

eiriktsarpalis commented on src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs:
```diff
@@ -0,0 +1,272 @@
+                        }
+                        else
+                        {
+                            // For each subsequent element in the source, randomly replace an element in the
```
Sampling then shuffling is an interesting approach. Do we know/care what impact this might have from a distribution perspective?

stephentoub replied:
Reservoir sampling produces a uniform distribution. A final shuttle then ensures the selected elements are in a random order, with every element having an equal chance of each position. I believe it's ok.

stephentoub replied:
cc: @bartonjs to check me on this

bartonjs replied:
I was afraid that the random sort was always going to be bounded on takeCount, which would make for a very skewed distribution.

Since the call is always to Next(i + 1) this looks the same as the ascending triangle implementation of Fisher-Yates (vs the "more common" (from what I've seen) descending triangle).

Since this implementation looks substantially the same as

```C#
byte[] copy = (byte[])source.Clone();
Shuffle(copy);
Array.Resize(ref copy, takeCount);
```

it feels uniform to me.  Running 10k iterations of Range(0, 1000).Shuffle().Take(10) and seeing what the hit counts look like would be telling.  All 1000 entries should show up with (roughly) equal occurrence over that range.  The test at https://github.com/dotnet/runtime/pull/111015#issuecomment-2569165761 might also be useful, though it probably requires an adjustment for only doing sub-sampling that isn't obvious off the top of my head.

stephentoub replied:
Thanks for reviewing. I've added a test.

eiriktsarpalis commented on src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs:
```diff
@@ -0,0 +1,272 @@
+                    return base.Take(count);
+                }
+
+                // Otherwise, we either don't know how many elements are in the source, or we know it's more than count.
```
My understanding is that builds of System.Linq that have this optimization enabled will behave differently when calling `Shuffle().Take()` on infinite enumerables. Are we ok with this?

eiriktsarpalis replied:
Correspondingly `infiniteEnumerable.Shuffle().Take(1).Count()` completes whereas `await infiniteEnumerable.AsAsyncEnumerable().Shuffle().Take(1).CountAsync()` diverges.

stephentoub replied:
We've now shifted away from two different builds and instead just have a single build with a feature switch, so it's really a question of whether the feature switch impacts behavior, and that's much more palatable, since that's what feature switches are for :)

This whole file doesn't come into play if the size-optimized feature switch is set, so I'm not sure there's much we can do about it, unless you're suggesting blocking enumerables > 2^31 elements from working even when we can support them?

bartonjs commented on src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs:
```diff
@@ -0,0 +1,272 @@
+                }
+                else if (index >= 0)
+                {
+                    TSource[] array = _source.ToArray();
```
How often is TryGetElementAt expected to be called? ToArray() on every index operation seems quite expensive to me

stephentoub replied:
It's no more expensive han if this wasn't here, in which case ElementAt which still do this. It's similar to Reverse().ElementAt on an enumerable today, or OrderBy(). ElementAt. Hopefully it's not frequently used. That said, now that I have the sample method, I can use that here to take away a bit of the sting, though it'll still be O(n).

bartonjs replied:
I think my implied question was "should this do `_source = _source.ToArray()` if this ever gets hit?", but perhaps not.

stephentoub replied:
> I think my implied question was "should this do `_source = _source.ToArray()` if this ever gets hit?", but perhaps not.

That would break the typical LINQ assumption that every iteration is fresh.

I've changed the implementation to not create the full array each time, though it does still need to fully enumerate.

bartonjs commented on src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs:
```diff
@@ -0,0 +1,272 @@
+            public override List<TSource> ToList() => SampleToList(_source, _takeCount) ?? [];
+
+            public override int GetCount(bool onlyIfCheap) =>
+                !onlyIfCheap ? Math.Min(_takeCount, _source.Count()) :
```
`_source` isn't a take-limited enumerator, is it?  Would we want this to be `Min(_takeCount, _source.Take(_takeCount).Count())`, or would under-enumerating there be a Linq side-effect no-no?

stephentoub replied:
Oh, interesting. The downside to that is allocating another iterator object for the Take. The upside is we could avoid a ton of enumeration. On the balance, it's probably worthwhile. I've made the change but swapped the order of the checks so we only take that path if we can't get the count O(1).

bartonjs commented on src/libraries/System.Linq/tests/ShuffleTests.cs:
```diff
@@ -0,0 +1,316 @@
+                    foreach (int i in source.Shuffle().Take(takeCount + 1).Take(takeCount)) second.Add(i);
+                    Assert.Equal(Math.Min(takeCount, length), first.Count);
+                    Assert.Equal(Math.Min(takeCount, length), second.Count);
+                    Assert.NotEqual(first, second);
```
Is it worth asserting that first.Distinct().Count() is also equal to the requested count?  That is, that none of "too short", "too long", (or, elsewhere "just right") has a path that duplicates entries?

bartonjs commented on src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs:
```diff
@@ -2494,7 +2494,7 @@ static HttpRequestHeaders CreateHeaders()
                 };
 
-                foreach (Action<HttpRequestHeaders> action in actions.OrderBy(_ => Random.Shared.Next()))
+                foreach (Action<HttpRequestHeaders> action in actions.Shuffle())
```
Hopefully this went into your analyzer notes for patterns to match against

bartonjs reviewed (approved):
The reservoir collection in the Shuffle-Take iterator took some thinking about, but it seems legit to me.

stephentoub committed:
Merge branch 'main' into addshuffle

stephentoub committed:
Improve impl and tests

- Avoid ToArray in TryGetElement, instead using SampleToList
- Rename TakeShufferIterator to ShuffleTakeIterator
- Avoid full enumeration in ShuffleTakeIterator.GetCount()
- Add distribution test

eiriktsarpalis commented on src/libraries/System.Linq.AsyncEnumerable/src/System/Linq/Shuffle.cs:
```diff
@@ -0,0 +1,59 @@
+#if NET
+                Random.Shared.Shuffle(array);
+#else
+                Random random = t_random ??= new Random(Environment.TickCount ^ Environment.CurrentManagedThreadId);
```
Not a huge deal, but couldn't this be expressed as an extension method we polyfill?

stephentoub replied:
Not until we have extension properties and extension statics.

A real polyfill for it would also be much more code, as the instance returned by Shared is usable from any thread, not just the current one.

ryanheath commented on src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs:
```diff
@@ -0,0 +1,280 @@
+                TryGetElementAt(0, out found);
+
+            public override TSource? TryGetLast(out bool found) =>
+                TryGetElementAt(0, out found);
```
@stephentoub not sure if TryGetLast is impl correctly here?

stephentoub replied:
It is. This is a shuffle, so the index doesn't matter; last is the same as first.
